src/dune_engine/scheduler.ml

```diff
@@ -893,7 +901,9 @@ let prepare (config : Config.t) ~polling =
               else
                 loop (Filename.concat acc "..") (Filename.dirname dir)
             in
-            loop ".." (Filename.dirname s) ) ) );
+            loop ".." (Filename.dirname s) ) )
+    in
+    Console.print [ Pp.verbatim (sprintf "Entering directory '%s'" dir) ] );
```

Comment:
I want to kick this print out of `Scheduler` as well, but I'm not sure how to make it work with the current `Handler`. One change that seems reasonable is to only print this message if we are building something. As far as I know, this `Entering directory` message is only useful for editors to parse build output, so I think it makes sense to only print if we call `go` or `poll`. @aalekseyev what do you think?

Reply:
I don't understand neither the problem, nor the proposal.
Can't you add a field to the `Handler` to print this?
What does it mean to "only" print in `go` or `poll`? Is there a third alternative?

Reply:
Oops, I think I confused myself. I was thinking that `prepare` was called separately from `go` or `poll`.

> Can't you add a field to the Handler to print this?

That's what I'm proposing.

> What does it mean to "only" print in go or poll? Is there a third alternative?

My idea was that there should be. There's 3 use cases for running the scheduler:
* Single build
* Polling build
* Running some misc command that doesn't require building. E.g. `$ dune subst`.

I thought that we need to skip the printing of cwd for `$ dune subst`, but we can already do it by setting `Clflags.no_print_directory`, so it doesn't matter.

Reply:
I pushed a commit for this. It requires us to add `on_event` to `Scheduler.go`. It's a bit of a shame that `on_event` for `Scheduler.go` has to handle events that can only happen when polling. Avoiding this was my original motivation for `Scheduler.Build.t`, but there seems to be no practical benefits apart from a little extra type safety.

Reply:
Splitting `Build.event` seems reasonable to me. Extra type safety is a very tangible benefit, also serving as a documentation.

By the way, from what I can tell, in the case of `dune subst` the asynchronous computation is not even a build, so a lot of this terminology seems wrong for a general case.

Reply:
> Splitting Build.event seems reasonable to me. Extra type safety is a very tangible benefit, also serving as a documentation.

Okay, let me see what I can do.

> By the way, from what I can tell, in the case of dune subst the asynchronous computation is not even a build, so a lot of this terminology seems wrong for a general case.

That's true. I like the term `Job.t` more, but it already means something in the context of a scheduler. Perhaps `Task.t`?

Reply:
Printing the "Entering directory ..." message is a one-off thing. There is no need to have an event for this. I removed the event and moved the printing to `Common.set_common`. It seems that this fixed a double printing problem in `dune exec`, so that's nice.

I also added a comment explaining why we print this message.